src/Text/Pandoc/Readers/Man.hs

```diff
@@ -0,0 +1,552 @@
+--
+
+data FontKind = Regular | Italic | Bold | ItalicBold deriving Show
+
+data MacroKind = KTitle
+               | KCodeBlStart
+               | KCodeBlEnd
+               | KTab
+               | KTabEnd
+               | KSubTab
+               deriving (Show, Eq)
+
+type RoffStr = (String, FontKind)
+
+data ManToken = MStr RoffStr
+              | MLine [RoffStr]
+              | MMaybeLink String
+              | MEmptyLine
+              | MHeader Int [RoffStr]
+              | MMacro MacroKind [RoffStr]
+              | MUnknownMacro String [RoffStr]
+              | MComment String
+              deriving Show
+
+data EscapeThing = EFont FontKind
+                 | EChar Char
+                 | ENothing
+                 deriving Show
+
+data RoffState = RoffState { fontKind :: FontKind
+                           } deriving Show
+
+instance Default RoffState where
+  def = RoffState {fontKind = Regular}
+
+type ManLexer m = ParserT [Char] RoffState m
+type ManParser m = ParserT [ManToken] ParserState m
+
+---- debug functions
+{-
+import Text.Pandoc.Class (runIOorExplode)
+
+printPandoc :: Pandoc -> [Char]
+printPandoc (Pandoc m content) =
+  let ttl = "Pandoc: " ++ (show $ unMeta m)
+      cnt = intercalate "\n" $ map show content
+  in ttl ++ "\n" ++ cnt
+
+testStr :: String -> IO ()
+testStr str = do
+  pand <- runIOorExplode $ readMan def (T.pack str)
+  putStrLn $ printPandoc pand
+
+  
+testFile :: FilePath -> IO ()
+testFile fname = do
+  cont <- readFile fname
+  testStr cont
+-}
+----
+
+
+-- | Read man (troff) from an input string and return a Pandoc document.
+readMan :: PandocMonad m => ReaderOptions -> T.Text -> m Pandoc
+readMan opts txt = do
+  eithertokens <- readWithM lexMan def (T.unpack $ crFilter txt)
+  case eithertokens of
+    Right tokenz -> do
+      let state = def {stateOptions = opts} :: ParserState
+      eitherdoc <- readWithMTokens parseMan state tokenz
+      case eitherdoc of
+        Right doc -> return doc
+        Left e -> throwError e
+    Left e       -> throwError e
+
+  where
+
+  readWithMTokens :: PandocMonad m
+          => ParserT [ManToken] ParserState m a  -- ^ parser
+          -> ParserState                         -- ^ initial state
+          -> [ManToken]                       -- ^ input
+          -> m (Either PandocError a)
+  readWithMTokens parser state input =
+    mapLeft (PandocParsecError . (intercalate "\n") $ show <$> input) `liftM` runParserT parser state "source" input
+
+  mapLeft :: (a -> c) -> Either a b -> Either c b
+  mapLeft f (Left x) = Left $ f x
+  mapLeft _ (Right r) = Right r
+
+--
+-- String -> ManToken function
+--
+
+lexMan :: PandocMonad m => ManLexer m [ManToken]
+lexMan = many (lexComment <|> lexMacro <|> lexLine <|> lexEmptyLine)
+
+parseMan :: PandocMonad m => ManParser m Pandoc
+parseMan = do
+  let parsers = [ try parseList, parseTitle, parsePara, parseSkippedContent
+                , try parseCodeBlock, parseHeader, parseSkipMacro]
```

Comment:
These parsers return a `Block`.  In the other pandoc readers, we return a `Blocks`, generated using the functions in Text.Pandoc.Builder.  (Ditto for `Inline`).  This is recommended for several reasons:  (a) it will make changes to the underlying types less painful to implement, (b) you automatically get things like merging of adjacent Space elements and elimination of Null blocks, (c) your parsers can return empty sequences (`mempty`) or sequences containing more than one block or inline.